SUR-223 - Search component prototype

HimanshuSangale opened the pull request:
### Description

The `SearchBox` component is a customizable search input component built with React and styled using Tailwind CSS. It provides a user-friendly interface for search functionality and supports dropdown results and loading states.

### Screenshots

https://d.pr/i/J3NUGV

### Checklist:

-   [x] My code is tested
-   [ ] My code passes the PHPCS tests
-   [x] I've created the npm build.
-   [ ] My code follows accessibility standards <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/accessibility-coding-standards/ -->
-   [x] My code has proper inline documentation <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/javascript/ -->
-   [ ] I've included any necessary tests <!-- if applicable -->
-   [ ] I've included developer documentation <!-- if applicable -->
-   [ ] I've added proper labels to this pull request <!-- if applicable -->

vrundakansara commented on src/components/search/search.jsx:
```diff
@@ -0,0 +1,83 @@
+	loadingIcon = {}, // Icon shown during loading
+	searchResult = ['searchResult 1', 'searchResult 2', 'searchResult 3'],
+	additionalResult = ['Category 1', 'Category 2', 'Category 3'],
+	searchResultIcon = < File size={16} strokeWidth={2} />,
+	additionalResultIcon = <Folder size={16} strokeWidth={2} />,
```
Sizes of the icons should be handled by the parent div - 
Ex: https://github.com/brainstormforce/force-ui/blob/master/src/components/button/button.jsx#L46

vrundakansara commented on src/components/search/search.jsx:
```diff
@@ -0,0 +1,83 @@
+	size = "sm", // sm, md, lg
+	loading = false,
+	loadingIcon = {}, // Icon shown during loading
+	searchResult = ['searchResult 1', 'searchResult 2', 'searchResult 3'],
+	additionalResult = ['Category 1', 'Category 2', 'Category 3'],
```
We don't really need these defaults

vrundakansara commented on dist/utils/withTW.js.map:
Run npm run build once to get rid of this extra file

vrundakansara commented on src/components/search/search.jsx:
```diff
@@ -0,0 +1,83 @@
+	return (
+		<div className={cn('relative flex flex-col focus-within:z-10', className)}>
+			<Input
+				prefix={<Search />}
```
We should be using input type search here - https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input/search

HimanshuSangale commented:
I will make the above suggested changes. @vrundakansara

BabuTechBsf commented:
Loading Shows this : https://d.pr/i/CTYlka
Loading Icon - https://d.pr/i/wUVkW7
Search results Icon - https://d.pr/i/sv1BsK and 
Additional Result Icon - check 
check how you can increase the height so that the results will be visible (storybook table change) - https://d.pr/v/D8PGxQ

BabuTechBsf commented on src/components/search/search.jsx:
```diff
@@ -0,0 +1,172 @@
+		loadingIcon = <Loader />, // Default loading icon
+		searchResult = [],
+		additionalResult = [],
+		searchResultIcon = <File />,
```
result icon set only if passed

BabuTechBsf commented on src/components/search/search.jsx:
```diff
@@ -0,0 +1,172 @@
+		searchResult = [],
+		additionalResult = [],
+		searchResultIcon = <File />,
+		additionalResultLabel = 'Category',
```
Custom text should not be placed. In this case set label if something is passed only.

BabuTechBsf commented on src/components/search/search.jsx:
```diff
@@ -0,0 +1,172 @@
+		additionalResult = [],
+		searchResultIcon = <File />,
+		additionalResultLabel = 'Category',
+		additionalResultIcon = <Folder />,
```
resultIcon set only if passed

BabuTechBsf commented on src/components/search/search.jsx:
```diff
@@ -0,0 +1,172 @@
+		children,
+	} = props;
+
+	const [ isOpen, setIsOpen ] = useState( false );
```
prefer varialble names associated with search, searchResultOpen etc..

BabuTechBsf commented on src/components/search/search.jsx:
```diff
@@ -0,0 +1,172 @@
+		};
+	}, [] );
+
+	const commonClass = 'relative w-full focus-within:z-10';
```
Move your static styles to separate file. Check radio component or select.

BabuTechBsf commented on src/components/search/search.jsx:
```diff
@@ -0,0 +1,172 @@
+		? 'border-border-disabled bg-field-background-disabled cursor-not-allowed text-text-disabled'
+		: '';
+
+	const handleSearchChange = ( e ) => {
```
handleSearch is enough i guess.

BabuTechBsf commented on src/components/search/search.jsx:
```diff
@@ -0,0 +1,172 @@
+				<span className={ cn( iconClasses, sizeClasses, 'left-0 z-20 text-gray-500 group-hover:text-text-primary' ) }>
+					<Search />
+				</span>
+				<input
```
Since input is root element, you can take args for this and extend here, like classname etc...

BabuTechBsf commented on src/components/search/search.jsx:
```diff
@@ -0,0 +1,172 @@
+									No Results found
+								</div>
+							) }
+							{ /* Additional categories */ }
```
additional results

BabuTechBsf reviewed (changes requested):
What is additionalResults here, Is it given for stack of results or two set only.
Please contact designer for information.

JelenaTakac commented on src/components/search/search.jsx:
```diff
@@ -0,0 +1,172 @@
+import { cn } from '@/utilities/functions';
+import { File, Folder, Search, SquareSlash } from 'lucide-react';
+import { useState, useEffect } from 'react';
+import { Label } from '..';
+import Loader from '../loader';
```
Please ensure consistency in how outside components are imported.

JelenaTakac commented on src/components/search/search.jsx:
```diff
@@ -0,0 +1,172 @@
+	const baseClasses = 'font-normal placeholder-text-tertiary text-text-secondary w-full focus:outline-none';
+
+	const sizeClasses = {
+		sm: 'p-1 rounded-md text-xs [&>svg]:h-4 [&>svg]:w-4',
+		md: 'p-1.5 rounded-md text-sm [&>svg]:h-5 [&>svg]:w-5',
+		lg: 'p-2 rounded-lg text-base [&>svg]:h-6 [&>svg]:w-6',
```
Instead of specifying both width `w` and height `h` for the SVG, you can use `[&>svg]:size` to apply the same dimensions uniformly.

jaieds reviewed (changes requested):
1. Showing horizontal scrollbar.
2. Remove the hover effect for the disabled state.
3. Add some transition on hover.
4. Fix glitch.
    - Issue Details: While searching/typing in the search box, it shows another result for a few seconds before showing the final result again. It should show a loading icon until the final result is ready.

**Screencast:** https://d.pr/v/kYENYq

jaieds commented on src/components/search/search.jsx:
```diff
@@ -0,0 +1,251 @@
+));
+SearchBoxSeparator.displayName = 'SearchBoxSeparator';
+
+export {
+	SearchBox,
+	SearchBoxInput,
+	SearchBoxContent,
+	SearchBoxLoading,
+	SearchBoxResults,
+	SearchBoxResultTitle,
+	SearchBoxResultItem,
+	SearchBoxSeparator,
+};
```
Please follow how we are exporting other components.

**Export:**
```jsx
export default SearchBox
```

**Import and usage:**
```jsx
import {SearchBox} from '@bsf/force-ui'
<SearchBox>
...
<SearchBox.Input />
...
</SearchBox>
```

HimanshuSangale replied:
Fixed!

jaieds commented on src/components/search/search.jsx:
```diff
@@ -0,0 +1,251 @@
+	return (<div
+		ref={ref}
+		className={cn(
+			'flex items-center gap-2 p-1 hover:bg-gray-200',
```
Use colors that are defined in the Tailwind Config/Figma.

❌ `hover:bg-gray-200`

HimanshuSangale replied:
Fixed!

ravindra114 reviewed (changes requested):
https://d.pr/i/2WH9ny

storybook code seems incorrect

ravindra114 commented on src/components/search/search.jsx:
```diff
@@ -0,0 +1,249 @@
+		<div
+			tabIndex={0}
+			ref={refs.setReference}
+			className={cn("w-full group relative flex justify-center items-center gap-2 focus-within:z-10",
```
use gap-0.5

HimanshuSangale replied:
Resolved this as per discussion, added custom gaps for better placements.

jaieds reviewed (changes requested):
I've found one issue and suggested some UI improvements. Please watch the video

👉 https://share.cleanshot.com/DLdZ3fJq

ravindra114 reviewed (commented):
can you please check all 'gaps' once?

jaieds reviewed (changes requested):
1. The border around the dropdown is missing.
2. Improvement: Show a message like `No results found` when there's no search result.
3. Optional improvement: Select options using keyboard arrow keys.

**Screenshot:** https://d.pr/i/suLfG8

HimanshuSangale commented:
- @jaieds I've implemented the visual changes as per your suggestions, including the hover transition we discussed.
- @ravindra114 I've maintained the original gap based on our conversation, as reducing it to 0.5 makes the layout appear cramped.

jaieds reviewed (changes requested):
@HimanshuSangale, There are some issues left. 

- Fix input font size for different sizes.
- Use `text-text-tertiary` for the `No results found` message.
- Rename the `dimension` prop to `size` for consistency.

Check the screenshot: https://d.pr/i/LXJ8wV

ravindra114 reviewed (changes requested):
Object.assign( elements.floating.style, {
							maxHeight: availableHeight,
})


instead of this you can directly assign
elements.floating.style.maxHeight = `${availableHeight}px`;

HimanshuSangale commented:
@jaieds 
- Input font size is already according to Figma design.
- Fixed color to text-text-tertiary for the No results found message. ✅ 
- Renames the dimension prop to size for consistency. ✅ 

@ravindra114 
- Refined the code for floating styles ✅

ravindra114 reviewed (changes requested):
https://d.pr/i/5jtQbD
can you check this once?

JelenaTakac commented on src/components/search/search.jsx:
```diff
@@ -0,0 +1,367 @@
+const SearchBox = forwardRef(
+	(
+		{
+			className, // Custom class names for additional styling
+			size = 'sm', // Size of the SearchBox (default is 'sm')
+			open = false, // Controls whether the dropdown is open or not (default is false)
+			onOpenChange = () => {}, // Callback function to handle the open state change
+			...props // Any additional props that are passed
```
Maybe you shouldn't leave comments here, since you explain the props in the readme file.

JelenaTakac commented on src/components/search/search.jsx:
```diff
@@ -0,0 +1,367 @@
+		return (
+			<SearchContext.Provider
+				value={ {
+					size, // Pass size to context
+					open, // Pass open state to context
+					onOpenChange, // Pass onOpenChange handler to context
+					refs, // Pass refs to context
+					floatingStyles, // Pass floating styles to context
+					context, // Pass context to context (nested contexts)
+					getReferenceProps, // Reference props for floating elements
+					getFloatingProps, // Floating props for dropdown content
```
Maybe you shouldn't leave comments here, since you explain the props in the readme file.

JelenaTakac commented on src/components/search/search.jsx:
```diff
@@ -0,0 +1,367 @@
+const SearchBoxInput = forwardRef(
+	(
+		{
+			className, // Custom class names for input styling
+			type = 'text', // Input type (default is 'text')
+			placeholder = 'Search...', // Placeholder text for input
+			variant = 'primary', // Input variant (default is 'primary')
+			disabled = false, // Whether input is disabled or not
+			value, // Controlled value for the input
+			defaultValue, // Default value if uncontrolled
+			onChange, // Callback for handling value changes
+			...props // Any additional props
```
Maybe you shouldn't leave comments here, since you explain the props in the readme file.

JelenaTakac commented on src/components/search/search.jsx:
```diff
@@ -0,0 +1,367 @@
+
+const SearchBoxContent = ( {
+	className,
+	dropdownPortalRoot = null, // Root element where the dropdown will be rendered.
+	dropdownPortalId = '', // Id of the dropdown portal where the dropdown will be rendered.
```
Maybe you shouldn't leave comments here, since you explain the props in the readme file.

HimanshuSangale commented:
@ravindra114 @jaieds Fixed separator width issue ✅ 
@JelenaTakac I've removed the comments ✅

BabuTechBsf reviewed (changes requested):
Search
https://d.pr/i/IWMThy
placeholder - on hover color
on hover - inner border color.
ghost - disabled - not match figma.

default value is not working on story docs
categories icon not showing up.
Few docs are not given:
Not sure how to pass class to root element.

ravindra114 reviewed (changes requested):
https://d.pr/i/PehluX

check storybook code? Rest things seems fine

HimanshuSangale commented:
@ravindra114 
- Made changes to demo code on storybook✅ 

@BabuTechBsf 
- Made changes to focus border color, hover colors shades look a little different but they match documentation given in Figma designs.✅ 
- removed controls for default value as it is part of uncontrolled usage.✅ 
- Category icons were intentionally hidden to show non-icon look, now added. ✅ 
- Ghost disable state fixed! ✅

BabuTechBsf reviewed (changes requested):
Search
Docs - don’t pass custom code for docs. you template should be rendered with the given args. and remove the unnecessary styles(use tailwind). The rendered code should look good to the user. He should be able to set all the args of his choice from story and use that code not static.
i don’t see rootClassname and wrapper classname.
We need to make all the args to be able to modify from story. value, default value also. refer other stories.

I have doubts regarding the implementation.
we are doing all the functionality(state management, logic etc..) and results UI in story level instead of compoenent level.
How do you manage Size in results are? is size only for input not results area?
My ideal would be use input and pass data object as prop that should show the result from that array. you can give result logic function to user.

on hover still the placeholder color is dark.

Create sub stories with opened state.

<File size={ 16 } />
For lucide icons we are following classname=size-4 like this pls follow that aproach.

HimanshuSangale commented:
@BabuTechBsf 
- Implemented dynamic code for documentation. ✅
- Added result filtering functionality directly within the component. Users can now pass items, and the component will handle the filtering and display logic. This can be overridden if custom filtering logic is required. ✅
- Updated the placeholder color to use the predefined placeholder color for hover and focus states, enhancing consistency with the UI. ✅
- Did not create a separate story for the open state, as it overlaps with other examples. However, the main story's arguments can be adjusted to control the dropdown's open and close states. ✅
- Added a new story to demonstrate the loading state. ✅

vrundakansara commented on src/components/search/styles.js:
```diff
@@ -0,0 +1,61 @@
+	ghost: 'bg-field-secondary-background border border-solid border-transparent',
+};
+
+export const IconClasses =
```
```suggestion
export const iconClasses =
```

vrundakansara commented on src/components/search/search.jsx:
```diff
@@ -0,0 +1,398 @@
+	variantClassNames,
+	disabledClassNames,
+	sizeClassNames,
+	IconClasses,
```
```suggestion
	iconClasses,
```

vrundakansara commented on src/components/search/search.jsx:
```diff
@@ -0,0 +1,398 @@
+				<span
+					className={ cn(
+						textSizeClassNames[ size ],
+						disabled ? 'text-icon-disabled' : IconClasses,
```
```suggestion
						disabled ? 'text-icon-disabled' : iconClasses,
```

vrundakansara commented on src/components/search/search.jsx:
```diff
@@ -0,0 +1,398 @@
+				<span
+					className={ cn(
+						textSizeClassNames[ size ],
+						disabled ? 'text-icon-disabled' : IconClasses,
```
```suggestion
						disabled ? 'text-icon-disabled' : iconClasses,
```

BabuTechBsf commented:
> Search Docs - don’t pass custom code for docs. you template should be rendered with the given args. and remove the unnecessary styles(use tailwind). The rendered code should look good to the user. He should be able to set all the args of his choice from story and use that code not static. i don’t see rootClassname and wrapper classname. We need to make all the args to be able to modify from story. value, default value also. refer other stories.
> 
> I have doubts regarding the implementation. we are doing all the functionality(state management, logic etc..) and results UI in story level instead of compoenent level. How do you manage Size in results are? is size only for input not results area? My ideal would be use input and pass data object as prop that should show the result from that array. you can give result logic function to user.
> 
> on hover still the placeholder color is dark.
> 
> Create sub stories with opened state.
> 
> <File size={ 16 } /> For lucide icons we are following classname=size-4 like this pls follow that aproach.

@HimanshuSangale Pls add your response to all the points.
Did not create a separate story for the open state - > You can use custom height to showcase open state. (story level)

HimanshuSangale commented:
> @BabuTechBsf
> 
> * Implemented dynamic code for documentation. ✅
> * Added result filtering functionality directly within the component. Users can now pass items, and the component will handle the filtering and display logic. This can be overridden if custom filtering logic is required. ✅
> * Updated the placeholder color to use the predefined placeholder color for hover and focus states, enhancing consistency with the UI. ✅
> * Did not create a separate story for the open state, as it overlaps with other examples. However, the main story's arguments can be adjusted to control the dropdown's open and close states. ✅
> * Added a new story to demonstrate the loading state. ✅

Last second point mentions about the open state, User can check from basic story using open state control. Added loading state story.
